### ISSUE TITLE
Idea for new Example: Non-uniform Tline

### ISSUE THREAD

**Issue by Jormit** [paramrf/paramrf#45]
I'm exploring a new idea for an example but don't know the best way to approach it.
The idea is to optimise a non-uniform transmission line, where the width of the line is defined by some function and you optimise the function parameters instead of the raw values. 

I'm thinking of breaking the line up into smaller segments

I have this right now:

```python

import pmrf
import matplotlib.pyplot as plt
import numpy as np

from my_models import non_uniform_tlin

tlin = non_uniform_tlin(
    10,
    lambda y: y * pmrf.Bounded(1.0, 2.0, value=1.0),
    w1=pmrf.Fixed(4e-3),
    w2=pmrf.Fixed(6e-3),
    h=pmrf.Fixed(2.0e-3),
    epr=pmrf.Fixed(4.6),
    tand=pmrf.Fixed(0.025),
    rho=pmrf.Fixed(1.72e-8),
    length=pmrf.Fixed(10e-3),
)


match_goal = pmrf.evaluators.Goal('s11_db', '<', -20)
freq = pmrf.Frequency(1, 50, 1001, "GHz")
result = pmrf.optimize.minimize(match_goal, tlin, freq, solver=pmrf.optimize.ScipyMinimize(method='L-BFGS-B'))
optimized_tlin = result.model

from pprint import pprint
print("Original Parameters:")
pprint(tlin.named_params())

print("Optimized Parameters:")
pprint(optimized_tlin.named_params())

plt.plot(freq.f, -20.0 * np.ones_like(freq.f), color='black', linestyle='--', label='target')
tlin.plot_s_db(freq, m=0, n=0, label='initial')
optimized_tlin.plot_s_db(freq, m=0, n=0, label='optimized')
plt.title('Initial vs. Optimized S11')

plt.show()
```

Line is defined in other file like this:
```python
def non_uniform_tlin(num_seg, func, w1, w2, h, epr, tand, rho, length):
    delta_l = length / num_seg
    lines = []
    for i in range(num_seg):
        ratio =  func((i + 0.5) / num_seg)
        w = w1 * (1 - ratio) + w2 * (ratio)

        lines.append(
            pmrf.models.MicrostripLine(
                length=delta_l, w=w, h=h, epr=epr, tand=tand, rho=rho, 
            )
        )

    tlin = pmrf.models.Cascade(lines)

    return tlin
```

This doesn't seem to work. What I want is to simply optimise this function:
```lambda y: y * pmrf.Bounded(1.0, 2.0, value=1.0)```
All the widths should be tied to this. But it doesn't seem to do that. 
In-fact it seems to even change the dimensions even if they are fixed.

Original parameters (linearly increasing widths):
```
{ 'cascade[0].w': 0.004100000000000002,
 'cascade[1].w': 0.004299999999999997,
 'cascade[2].w': 0.004500000000000003,
 'cascade[3].w': 0.004699999999999997,
 'cascade[4].w': 0.0049000000000000024,
 'cascade[5].w': 0.0051,
 'cascade[6].w': 0.005300000000000001,
 'cascade[7].w': 0.005499999999999996,
 'cascade[8].w': 0.005700000000000002,
 'cascade[9].w': 0.005899999999999996}
```

Optimised parameters:
```
{ 'cascade[0].w': 0.003479769872513286,
 'cascade[1].w': 0.11579579044179113,
 'cascade[2].w': 0.0024281190595549043,
 'cascade[3].w': 0.2286494737146181,
 'cascade[4].w': 0.005020850098384795,
 'cascade[5].w': 0.005184977045248878,
 'cascade[6].w': 0.20557010624589073,
 'cascade[7].w': 0.0055327203010588485,
 'cascade[8].w': 0.005761536311430941,
 'cascade[9].w': 0.006327246967435763}
```

Is there some way to do this properly? 
Maybe I could add some dummy model and tie() the parameters to that or something?
Do you have any suggestions for me?

**Comment by gvcallen** [comment]
Thanks for the suggestion.

I definitely want to formally support NUTLs in ParamRF, so having an example to demonstrate that is a definitely good idea.

Your method doesnt work because parameters cast to arrays as soon as you try to perform any operation on them. So e.g. `2.0 * pmrf.Bounded(10.0, 20.0, value=5.0)` will produce an array of `30.0`. This is by design - parameters should be treated as independent variables from which other values are derived.

If you wanted to make your example work, your lambda can just be modified to `lambda y: pmrf.Bounded(y*1.0, y*2.0, value=y*1.0)` and that should successfully constraint the parameters.

I do, however, want to get generic NUTL classes into the API too. Since this is a classic example for a re-usable component, it should definitely be a class, so I do recommend looking at the example docs for the syntax for that, since if we published an example I would prefer it to be class-based.

I started an implementation a while ago as an experiment, but I didn't really have a need for it personally, and then ParamRF's API changed, so it is currently commented out. If you are interested, you can have a look in `pmrf/models/components/lines/nonuniform.py`.

Edit: Also, for a truly general NUTL class in base ParamRF, the class should take a `solver` parameter to actual compute it. Have a look at e.g. `CoaxialLine` as a simple example.

**Comment by gvcallen** [comment]
Hi @Jormit , it's been a while

ParamRF v0.32 now supports a much more flexible approach to defining parameters. All of the optimize and fitting functions now accept arbitrary JAX "PyTrees" as inputs. This includes Python dictionaries and the new `pmrf.Module` class. Essentially this means you can have one, consistent, flat representation for all your independent parameters, and use that to construct models in whatever way you like, while the complete transmission line is rebuilt from it whenever the objective is evaluated:

```python
  class NUTLParams(pmrf.Module):
      scale: pmrf.Param

  params = NUTLParams(
      scale=pmrf.Bounded(1.0, 2.0, value=1.0),
  )

  def build_tlin(params):
      return non_uniform_tlin(
          10,
          lambda y: y * params.scale,
          w1=4e-3,
          w2=6e-3,
          h=2e-3,
          epr=4.6,
          tand=0.025,
          rho=1.72e-8,
          length=10e-3,
      )

  match_goal = pmrf.evaluators.Goal("s11_db", "<", -20)
  freq = pmrf.Frequency(1, 50, 1001, "GHz")

  result = pmrf.optimize.minimize(
      lambda params, freq: match_goal(build_tlin(params), freq),
      params,
      freq,
      solver=pmrf.optimize.ScipyMinimize(method="L-BFGS-B"),
  )

  optimized_tlin = build_tlin(result.model)
```

Here, scale is the only independent optimization parameter. All segment widths are recalculated from it, so they cannot move independently. You can also use a simple dictionary:

```python
  params = {
      "scale": pmrf.Bounded(1.0, 2.0, value=1.0),
  }
```

You would then replace params.scale with params["scale"] in build_tlin. The Module version is mainly useful when the parameter set is reused or becomes more complicated.

**Comment by Jormit** [comment]
Hi @gvcallen 

Thanks for the update. I haven't managed to look into this further yet. 

To get it clear in my head:
Is it correct to say that the pmrf.module class is only if you want to have a parameter container that could be seperate from the actual pmrf.model? 

So the main use case is to have multiple pmrf.models that accept the same parameter set but have different underlying implementations?

For my own things will it be fine into the future to use the pmrf.model like before, except for the deprecated build() method?

**Comment by gvcallen** [comment]
Yes, `pmrf.Module` is like a container for parameters, almost like a set of global variables like you'd have in AWR. However it is also more flexible than that, as you can also contain `pmrf.Model` instances and even other modules inside it. Anything inside the module tree is potentially a free parameter, whereas any model built from it e.g. in a cost function is not seen by the optimizer. There is a new example in the docs if you are interested

And yes, `pmrf.Model` should work just as before, except for the `build()` method. It was a tough choice to deprecate `build()` as I was using it myself, but I realized having a cleaner separation of "provides RF response" vs just "is a parameter container" was better. It should still work though, as I need to experiment with the new pattern myself to be sure, though this is my hunch.

**Comment by activexray** [comment]
perhaps of interest to you is my paper discussing AD optimization of NTLs https://ieeexplore.ieee.org/document/11021605

I actually never found a decent low-dimensional parameterization that worked better than just solving a discretized shape directly.

**Comment by gvcallen** [comment]
@activexray welcome!

did you every try iterative solving the Telegrapher ODEs directly? The JAX ecosystem makes this viable using the `diffrax` library, I think.

I briefly explored it myself but didn't really have the time

**Comment by activexray** [comment]
I would have if I had an equation of Z(x) 😅. I poked around at splines and like some polynomials but it's very difficult to add physical constraints to that model. The cascaded ABCD approach has worked very well for me and has excellent perf in reverse-mode.

**Comment by activexray** [comment]
Of course I could just also pick some x discretization like I do in the cascaded ABCD model and solve for the set of Z(x) at every x, but that seemed more complicated than just using the ABCD model. Plus, getting loss in there was tricky and was the whole point for the LNA matching problem.

**Comment by gvcallen** [comment]
`RepeatedCascade` (#176) has landed, and it covers the mechanism this issue ran into. Reporting here on how much of the issue it closes.

## What it fixes

The failure reported here was that `Cascade([MicrostripLine(w=...) for i in range(10)])` gives every section its own independent parameter — `cascade[0].w` ... `cascade[9].w` — so the optimiser moved ten free widths rather than the one coefficient in `lambda y: y * prf.Bounded(1.0, 2.0)`. That is exactly the optimised output pasted in the issue: ten unrelated widths, some of them 0.2 m.

`RepeatedCascade` replaces the list comprehension with one member model plus a mapping from its parameter names to arrays whose leading axis is the repeat axis:

```python
import jax.numpy as jnp
import pmrf as prf
from pmrf.models import MicrostripLine, RepeatedCascade

n = 10
line = MicrostripLine(w=4e-3, h=2e-3, length=10e-3 / n, ...)
tlin = RepeatedCascade(line, {'w': widths})   # widths.shape == (10,)
```

Concretely, against the three things this issue asked for:

- **One name set, not one per section.** The member is traced once and named once: `model.w`, `model.length`, `model.substrate.h`, ... There is no `cascade[k].w`. The per-repeat values are a single parameter, `values.w`.
- **Fixed parameters stay fixed.** The issue observed dimensions moving even when declared `Fixed`. Parameters not named in the mapping are genuinely one shared parameter for all repeats, so a `Fixed` one is one fixed parameter. The parameter that *is* repeated has its value on the member discarded, so `RepeatedCascade` fixes it at construction: it is not in the free set and an optimiser is never handed a parameter that moves nothing.
- **One trace.** Sections are built under a single `jax.vmap` over the repeat axis, not a Python loop, and reduced with the `pmrf.rf` cascade routines shared with `Cascade` (#175). The repeat count is static, taken from the leading axis of the supplied arrays.

S-parameters match a hand-built `Cascade` of the equivalent independently-parametrised members to ~1e-15, and gradients flow to each repeat individually, so the values can be computed from whatever you like — including a shape function — and fitted through.

The repeat axis is deliberately not the `prf.sweep` / `prf.batch_axes` parameter batch axis; a `RepeatedCascade` can carry both. That distinction is why it is not called `BatchedCascade`, and it is documented on the class. See ADR-0004.

## What remains

The *ergonomic* half of this issue is not closed. `RepeatedCascade` takes an array of widths; it does not take the function that produced them. To fit `lambda y: y * prf.Bounded(1.0, 2.0)` you would still evaluate that yourself at the section midpoints and hand the resulting array in — differentiable and correct, but you are writing the midpoint sampling and the `w1 (1 - ratio) + w2 ratio` blend by hand each time, and the fitted coefficient has no name or prior of its own inside the model.

That part belongs to `ProfiledLine` (ADR-0004), which is the intended answer for "optimise the function parameters instead of the raw values": a uniform line plus a mapping from parameter targets to `AbstractProfile` objects, whose coefficients are ordinary `prf.param` fields with units, constraints and priors. It samples the profiles at the section midpoints and builds a `RepeatedCascade`, so it inherits everything above. It also adds what this issue does not yet ask for but needs — the section count is a real accuracy knob, and an under-resolved taper is something an optimiser will exploit, so `ProfiledLine` carries Richardson extrapolation and a per-section electrical-length guard.

Two smaller things also remain for the example this issue is really about:

- The library ships `LinearProfile`, `ExponentialProfile` and `KlopfensteinProfile` only once `ProfiledLine` lands; a general fitted shape (polynomial, spline) is deferred to its own ticket per ADR-0004.
- The non-uniform-line example itself — the deliverable this issue opened with — is still to be written, on top of `ProfiledLine` rather than on `RepeatedCascade` directly.

So: the batching and naming defect reported here is fixed and testable today; the "optimise a function, not the values" interface and the example are `ProfiledLine`'s. Suggest keeping this open against the example, since that is what the title asks for.
